File: src/main/scala/com/vc/medium/L565.scala
package com.vc.medium

object L565 {
  def arrayNesting(arr: Array[Int]): Int = {
    var i = 0
    var j = 0
    var max = 0
    var size = 0
    val n = arr.length
    while(i < n) {
      size = 0
      j = i
      while(arr(j) >= 0) {
        size += 1
        val aj = arr(j)
        arr(j) = -1
        j = aj
      }
      max = Math.max(max, size)
      i += 1
    }
    max
  }
}

File: src/main/java/com/vc/medium/L565.java
package com.vc.medium;

class L565 {
    public int arrayNesting(int[] arr) {
        if(arr == null || arr.length == 0) return 0;

        int n = arr.length, index = 0, max = 0;
        for(int i = 0; i < n; i++) {
            if(arr[i] != -1) {
                int count = nesting(arr, i);
                max = Math.max(max, count);
            }
        }
        return max;
    }

    private int nesting(int[] arr, int index) {
        int count = 0;
        while(arr[index] != -1) {
            int oldIndex = index;
            index = arr[index];
            arr[oldIndex] = -1;
            count++;
        }
        return count;
    }
}
